src/IndexMaps/realindexmap.jl

```diff
@@ -75,12 +76,115 @@ function calculate_ind_values(imap::RealIndexMap, xs::Vector, dims::Vector{Int})
 end
 
-function grid_points(imap::RealIndexMap, N::Int, d::Int)
+""" 
+    grid_points(imap, N, d, span; exact_grid=true, enforced=[])
+
+  Gives `N` grid points from a given dimension of `imap` within a specified range.
+
+  # Arguments
+  - `imap::RealIndexMap`: An IndexMap specifying the structure of the TN being used
+  - `N::Int`: The number of grid points requested.
+  - `d::Int` The index of the dimension of `imap` requested
+  - `span::AbstractVector{<:Number}`: A two element number vector [a,b] with 0≤a<b≤1. The right endpoint of this span is not included as a gridpoint in the output.
+  - `exact_grid::Bool`: Flag specifying whether the function should give exact grid points 
+  (note: using `exact_grid=true` may cause less than `N` grid points to be returned)
+  - `enforced::AbstractVector{<:Number}`: A list of points that we want to enforce to show up in the grid.
+"""
+function grid_points(
+  imap::RealIndexMap,
+  N::Int,
+  d::Int;
+  span::AbstractVector{<:Number}=[0, 1],
+  exact_grid::Bool=true,
+  enforced=[],
+)
+  if length(span) != 2 || span[1] >= span[2] || span[1] < 0 || span[2] > 1
+    throw(
+      "expected a two-element vector [a,b] with 0≤a<b≤1 as input span, instead found $span"
+    )
+  end
+
+  dims = dim.(dimension_inds(imap, d))
+  @assert all(y -> y == first(dims), dims)
+  base = first(dims)
+  L = length(dimension_inds(imap, d))
+
+  if !exact_grid
+    grid_points = range(span[1], span[2] - (span[2] - span[1]) / N; length=N)
+  else #exact_grid = true
+    if span[1] == 0 && span[2] == 1
+      inv_step = min(Int(floor(log(base, N))), L)
+      if !(log(base, N) ≈ inv_step)
+        @warn "rounding $N down to $(Int(base^inv_step)) exact gridpoints!"
+      end
+      # now 1/inv_step should provide the step
+      grid_points = collect(span[1]:(1 / base^inv_step):span[2])[1:(end - 1)]
+
+    else
+
+      #get a count of the number of gridpoints within the span
+      points_in_span = floor(span[end] * base^L) - ceil(span[1] * base^L) + 1
+      #TODO: figure out a way to calculate this without exponentiating to the power of L
+
+      #exclude the endpoint in the count
+      if floor(span[end] * base^L) == span[end] * base^L
+        points_in_span = points_in_span - 1
+      end
+
+      if points_in_span <= 0
+        @warn "No exact gridpoints found in this span!"
+        grid_points = []
+      else
+        oldN = N
+        stepsize = ceil(points_in_span / N)
+        N = Int(floor(points_in_span / stepsize)) + 1
+        startval = ceil(span[1] * base^L) / base^L #startval is the smallest grid point ≥ span[1]
+
+        if startval + (N - 1) * stepsize / base^L >= span[2]
+          N = N - 1
+        end
+
+        if N < oldN
+          @warn "rounding $oldN down to $N exact gridpoints to maintain equal spacing."
+        end
+
+        grid_points = [startval + i * (stepsize / base^L) for i in 0:(N - 1)]
+      end
+    end
+  end
+
+  #add enforced points in naively
```

Comment:
Why not just
```
if !isempty(enforced)
  grid_points = sort(vcat(grid_points, enforced_points))` 
end
```
for this logic? I.e. just add them to the end of grid points and re-sort.
I also don't think we need to check the enforced points are in the span: that's something the user can ensure themselves for now